@vates/generator-toolbox/src/synchronized.mts

```diff
@@ -0,0 +1,145 @@
+      this.#nextValueForksReady = undefined
+      forksWaitingResolve() // for the other forks waiting
+    }
+    return promise
+  }
+
+  async next(uid: string): Promise<IteratorResult<T>> {
+    if (this.#removedForks.has(uid)) {
+      return { done: true, value: undefined }
+    }
+    if (!this.#forks.has(uid)) {
+      throw new Error(`trying to advance fork ${uid} that is not a fork of this one`)
+    }
+
+    if (this.#waitingForks.has(uid)) {
+      throw new Error(`Fork ${uid} is already waiting`)
+    }
+
+    this.#started = true
+    if (this.#nextValueForksReady === undefined) {
+      let forksWaitingResolve = () => {}
+      let forksWaitingReject: (reason?: Error) => void = () => {}
+      const next = this.#source.next().catch(async error => {
+        const e = new Error(`Error in the source generator ${error.message}`, { cause: error })
+        forksWaitingReject(e)
+        // source has failed, kill everything, and stop the forks
+        for (const uid of [...this.#forks.keys()]) {
+          await this.remove(uid, error)
+        }
+      })
+      const promise = Promise.all([
+        next,
+        new Promise((_resolve, _reject) => {
+          forksWaitingResolve = () => _resolve(undefined)
+          forksWaitingReject = _reject
+        }),
+      ]).then(([_]) => _ as IteratorResult<T>)
+
+      this.#nextValueForksReady = { promise, forksWaitingResolve, forksWaitingReject }
+    }
+    this.#waitingForks.add(uid)
+    return this.#resolveWhenAllForksReady()
+  }
+
+  async remove(uid: string, error?: Error): Promise<IteratorResult<T>> {
+    const fork = this.#forks.get(uid)
+    if (fork === undefined) {
+      if (this.#removedForks.has(uid)) {
+        // already removed
+        return { done: true, value: undefined }
+      }
+      throw new Error(`trying to remove fork wih uid ${uid} that is not a fork of this one`)
+    }
+    this.#forks.delete(uid)
+    this.#waitingForks.delete(uid)
+    this.#removedForks.add(uid)
+    try {
+      if (error === undefined) {
+        await fork.return()
+      } else {
+        await fork.throw(error)
+      }
+    } catch (cleaningError) {
+      console.error('Error while cleaning the forked', {
+        cleaningError,
+        sourceError: error,
+      })
+    }
+
+    if (this.#forks.size === 0) {
+      if (error === undefined) {
+        await this.#source.return(undefined as TReturn)
+      } else {
+        await this.#source.throw(error)
+      }
+      // Reject any pending forks waiting for the next value
+      if (this.#nextValueForksReady) {
+        this.#nextValueForksReady.forksWaitingReject(new Error('Source generator terminated.', { cause: error }))
+        this.#nextValueForksReady = undefined
+      }
+    } else {
+      // this fork was maybe blocking the others
+      if (this.#nextValueForksReady) {
+        await this.#resolveWhenAllForksReady()
+      }
+    }
+    return { done: true, value: undefined }
+  }
```

Comment:
When we terminate the `source`, don't you think we should clean up the `removedForks` to avoid keeping them in memory?
even if we just store a simple string, it may be better not to keep them in memory if we don't need them anymore

Reply:
it is a good practice to be strict in cleaning.